testutil/addr.go

```diff
@@ -1,23 +1,76 @@
 package testutil
 
 import (
+	"container/list"
+	"fmt"
 	"net"
+	"os"
 	"sync"
+	"time"
 )
 
-var bindLock sync.Mutex
-var bindNum byte = 10
+var (
```

Comment:
The changes here are an `ip` variation of the thing I recently fixed in consul's `sdk/freeport` library.